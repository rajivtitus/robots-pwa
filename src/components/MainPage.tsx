import { Component } from "react";

import CardList from "./CardList";
import SearchBox from "./SearchBox";
import Scroll from "./Scroll";
import ErrorBoundry from "./ErrorBoundry";
import Header from "./Header";
import { Robot } from "../types";

interface Props {
	onRequestRobots: () => any;
	onSearchChange: () => any;
	robots: Array<Robot>;
	searchField: string;
	isPending: boolean;
}

class MainPage extends Component<Props> {
	componentDidMount() {
		this.props.onRequestRobots();
	}

	filterRobots = (robots: Robot[]) => {
		return robots.filter((robot) => {
			return robot.name
				.toLowerCase()
				.includes(this.props.searchField.toLowerCase());
		});
	};

	render() {
		const { robots, onSearchChange, isPending } = this.props;
		const filteredRobots = this.filterRobots(robots);

		return (
			<div className="tc">
				<Header />
				<SearchBox searchChange={onSearchChange} />
				<Scroll>
					{isPending ? (
						<h1>Loading</h1>
					) : (
						<ErrorBoundry>
							<CardList robots={filteredRobots} />
						</ErrorBoundry>
					)}
				</Scroll>
			</div>
		);
	}
}

export default MainPage;
