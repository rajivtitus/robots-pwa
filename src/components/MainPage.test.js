import { shallow } from "enzyme";

import MainPage from "./MainPage";

let wrapper;

const mockRobots = [
	{
		id: 1,
		name: "Leanne Graham",
		username: "Bret",
		email: "Sincere@april.biz",
	},
];

beforeEach(() => {
	const mockStore = {
		onSearchChange: jest.fn(),
		onRequestRobots: jest.fn(),
		robots: [],
		searchField: "",
		isPending: false,
	};
	wrapper = shallow(<MainPage {...mockStore} />);
});

it("expects to render MainPage", () => {
	expect(wrapper).toMatchSnapshot();
});

it("filters robots correctly", () => {
	const mockProps = {
		onSearchChange: jest.fn(),
		onRequestRobots: jest.fn(),
		robots: mockRobots,
		searchField: "an",
		isPending: false,
	};

	const wrapper2 = shallow(<MainPage {...mockProps} />);

	expect(wrapper.instance().filterRobots([])).toEqual([]);
	expect(wrapper2.instance().filterRobots(mockRobots)).toEqual(mockRobots);
});
