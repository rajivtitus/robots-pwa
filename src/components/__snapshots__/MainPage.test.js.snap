// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expects to render MainPage 1`] = `
<div
  className="tc"
>
  <Header />
  <SearchBox
    searchChange={[MockFunction]}
  />
  <Scroll>
    <ErrorBoundary>
      <CardList
        robots={Array []}
      />
    </ErrorBoundary>
  </Scroll>
</div>
`;
